src/mg_decoration.cpp

```diff
@@ -169,33 +171,78 @@ size_t Decoration::placeDeco(Mapgen *mg, u32 blockseed, v3s16 nmin, v3s16 nmax)
 		for (u32 i = 0; i < deco_count; i++) {
 			s16 x = ps.range(p2d_min.X, p2d_max.X);
 			s16 z = ps.range(p2d_min.Y, p2d_max.Y);
-
 			int mapindex = carea_size * (z - nmin.Z) + (x - nmin.X);
 
-			s16 y = -MAX_MAP_GENERATION_LIMIT;
-			if (flags & DECO_LIQUID_SURFACE)
-				y = mg->findLiquidSurface(v2s16(x, z), nmin.Y, nmax.Y);
-			else if (mg->heightmap)
-				y = mg->heightmap[mapindex];
-			else
-				y = mg->findGroundLevel(v2s16(x, z), nmin.Y, nmax.Y);
-
-			if (y < y_min || y > y_max || y < nmin.Y || y > nmax.Y)
-				continue;
+			if ((flags & DECO_ALL_FLOORS) ||
+					(flags & DECO_ALL_CEILINGS)) {
+				// All-surfaces decorations
+				// Check biome of column
+				if (mg->biomemap && !biomes.empty()) {
+					std::unordered_set<u8>::const_iterator iter =
+						biomes.find(mg->biomemap[mapindex]);
+					if (iter == biomes.end())
+						continue;
+				}
+
+				// Get all floors and ceilings in node column
+				s16 floors[1024];
+				s16 ceilings[1024];
```

Comment:
Why 1024? The array size should be set, depending on the block size.
`(ymax - ymin) / 2` would be the maximal size of the array, if I'm not mistaken.

Reply:
Because `placeAllDecos()` can be called from `minetest.generate_decorations` which applies decorations to a lua voxelmanip of any size, i chose 1024 to be plenty big enough.

Reply:
Ah but we can get the size from nmax - nmin, ok.

Reply:
Tried but can't get it to work. When i replace 1024 with a variable both here and in `getSurfaces()` i get lots of errors, how should the arguments be in `getSurfaces()`?

Reply:
What prevents you from passing pointers as arguments for these functions?

Reply:
Ok i know that's an alternative method, i'll try that.

Reply:
Done.